pkg/cluster/enableendpoints_test.go

```diff
@@ -0,0 +1,295 @@
+package cluster
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	mgmtnetwork "github.com/Azure/azure-sdk-for-go/services/network/mgmt/2020-08-01/network"
+	"github.com/Azure/go-autorest/autorest/to"
+	"github.com/golang/mock/gomock"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	mock_subnet "github.com/Azure/ARO-RP/pkg/util/mocks/subnet"
+)
+
+var (
+	subscriptionId    = "0000000-0000-0000-0000-000000000000"
+	vnetResourceGroup = "vnet-rg"
+	vnetName          = "vnet"
+	subnetNameWorker  = "worker"
+	subnetNameMaster  = "master"
+	subnetIdWorker    = "/subscriptions/" + subscriptionId + "/resourceGroups/" + vnetResourceGroup + "/providers/Microsoft.Network/virtualNetworks/" + vnetName + "/subnets/" + subnetNameWorker
+	subnetIdMaster    = "/subscriptions/" + subscriptionId + "/resourceGroups/" + vnetResourceGroup + "/providers/Microsoft.Network/virtualNetworks/" + vnetName + "/subnets/" + subnetNameMaster
+)
+
+func TestEnableServiceEndpoints(t *testing.T) {
+	ctx := context.Background()
+
+	type test struct {
+		name          string
+		oc            *api.OpenShiftCluster
+		setMocks      func(subnetManagerMock *mock_subnet.MockManager, subnetsUpdaterMock *mock_subnet.MockEndpointsAdder, tt test)
+		expectedError string
+	}
+
+	tt := []test{
+		{
+			name: "should not call to subnetManager.CreateOrUpdate() because all endpoints are already in the subnet",
+			oc: &api.OpenShiftCluster{
+				Properties: api.OpenShiftClusterProperties{
+					MasterProfile: api.MasterProfile{SubnetID: subnetIdMaster},
+				},
+			},
+			setMocks: func(subnetManagerMock *mock_subnet.MockManager, subnetsUpdaterMock *mock_subnet.MockEndpointsAdder, tt test) {
+				subnet := &mgmtnetwork.Subnet{
+					ID: to.StringPtr(subnetIdMaster),
+					SubnetPropertiesFormat: &mgmtnetwork.SubnetPropertiesFormat{
+						ServiceEndpoints: &[]mgmtnetwork.ServiceEndpointPropertiesFormat{
+							{
+								Service:           to.StringPtr("Microsoft.ContainerRegistry"),
+								Locations:         &[]string{"*"},
+								ProvisioningState: mgmtnetwork.Succeeded,
+							},
+							{
+								Service:           to.StringPtr("Microsoft.Storage"),
+								Locations:         &[]string{"*"},
+								ProvisioningState: mgmtnetwork.Succeeded,
+							},
+						},
+					},
+				}
+
+				subnetIds := []string{tt.oc.Properties.MasterProfile.SubnetID}
+				endpoints := []string{"Microsoft.ContainerRegistry", "Microsoft.Storage"}
+				subnets := []*mgmtnetwork.Subnet{subnet}
+
+				subnetManagerMock.
+					EXPECT().
+					GetAll(gomock.Any(), subnetIds).
+					Return(subnets, nil)
+
+				subnetsUpdaterMock.
+					EXPECT().
+					AddEndpointsToSubnets(endpoints, subnets).
+					Return(nil)
+
+				subnetManagerMock.
+					EXPECT().
+					CreateOrUpdate(gomock.Any(), gomock.Any(), gomock.Any()).
+					Times(0)
+			},
```

Comment:
This doesn't follow DRY. All of these functions do pretty much the same thing but they are just copy pasted and modified a little bit. It would be better to have a single function to which we pass arguments.

Reply:
I do understand what you mean but this is just done this way to prepare the expectations of the mocks. As every test case is different, I don't see how we could create an "enough general" function that can be used to prepare the mocks for all the test cases. 🤔

Reply:
either find the parameters that allow you to have a general function, either split your tests into more specialized one. setup functions like these do not belong in the test parameters

Reply:
I have split the single Table Driven Test into multiple tests. Done.